### PR TITLE
Feature/project level metrics

### PR DESCRIPTION
**Pull request by nh13**
Will need to be rebased onto #30 so that the per-project TSV can have a filename prefix

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -309,9 +310,9 @@ sg001:17:A30ZZ:1:4:1234:4567:ACCTAG+TCCTGG 1:N:1:ACGT+TTGA
 
 Up to four metrics files are generated to help assess run and demultiplexing quality:
 
-##### `per_sample_metrics.tsv` 
+##### `per_sample_metrics.tsv`
 
-This file is always produced and contains the following columns:
+These two files are always produced and contains the following columns:
```
Comment (tfenne) [inline README.md:315]: This feels like maybe a change you meant to revert?  Which "two" files?

File: README.md [hunk 2]
```diff
@@ -309,9 +310,9 @@ sg001:17:A30ZZ:1:4:1234:4567:ACCTAG+TCCTGG 1:N:1:ACGT+TTGA
 
 Up to four metrics files are generated to help assess run and demultiplexing quality:
```
Comment (tfenne) [inline README.md:311]:
```suggestion
Up to five metrics files are generated to help assess run and demultiplexing quality:
```

File: README.md [hunk 3]
```diff
@@ -330,6 +331,16 @@ This file is always produced and contains the following columns:
 |`frac_q30_bases`|The fraction of bases in a template with a quality score 30 or above.|
 |`mean_index_base_quality`|The mean quality of index bases.|
 
+The `per_sample_metrics.tsv` file produces a row per sample.
+
+##### `per_project_metrics.tsv`
+
+The `per_project_metrics.tsv` file aggregates the metrics by project (aggregates the metrics across
+samples with the same project) and has the same columns as [per_sample_metrics.tsv](#per_sample_metricstsv).
+In this case, the `barcode_name` and `library_name` will contain the project name (or `None` if no
```
Comment (tfenne) [inline README.md:340]: What about the `barcode` field?  Same deal?

File: README.md (hunk 1 above)
Comment (nh13) [inline README.md:315]: fixme

File: src/lib/metrics.rs [hunk 4]
```diff
@@ -204,6 +204,94 @@ pub struct DemuxedGroupMetrics<'a> {
     pub sample_barcode_hop_tracker: Option<SampleBarcodeHopTracker<'a>>,
 }
 
+/// Write the per sample metrics.
+fn write_per_group_metrics<P: AsRef<Path>>(
+    samples: &[SampleMetadata],
+    total_templates: usize,
```
Comment (tfenne) [inline src/lib/metrics.rs:210]: Do you need this?  You could just total the per_group_metrics?

File: src/lib/metrics.rs [hunk 5]
```diff
@@ -204,6 +204,94 @@ pub struct DemuxedGroupMetrics<'a> {
...
+/// new sample corresponds to a unique project, and the associated metric are the sum of values
+/// for samples with the same project.  The undetermined sample is not grouped with any other
+/// sample.
+fn build_per_project_metrics(
+    samples: &[SampleMetadata],
+    per_sample_metrics: &[DemuxedGroupSampleMetrics],
+) -> Result<(Vec<SampleMetadata>, Vec<DemuxedGroupSampleMetrics>)> {
+    let mut project_to_metric: AHashMap<Option<&String>, DemuxedGroupSampleMetrics> =
+        AHashMap::new();
+    let mut project_to_sample: AHashMap<Option<&String>, SampleMetadata> = AHashMap::new();
```
Comment (tfenne) [inline src/lib/metrics.rs:247]: You can just use the stock HashMap here ... AHashMap is useful in performance sensitive contexts, but not here.

File: src/lib/metrics.rs [hunk 6]
```diff
@@ -204,6 +204,94 @@ pub struct DemuxedGroupMetrics<'a> {
...
+                let sample = SampleMetadata::new_allow_invalid_bases(
+                    project,
+                    BString::from(vec![b'N'; samples[0].barcode.len()]),
+                    project_to_metric.len(),
+                )?;
+                // Add the sample and current metric to the maps
+                project_to_sample.insert(key, sample);
+                project_to_metric.insert(key, metric.clone());
+            }
+            Some(project_metric) => project_metric.update_with_self(metric.clone()),
```
Comment (tfenne) [inline src/lib/metrics.rs:268]: `update_with_self` might just want to be `update_with` now?

File: src/lib/metrics.rs [hunk 7]
```diff
@@ -204,6 +204,94 @@ pub struct DemuxedGroupMetrics<'a> {
...
+    }
+
+    let mut project_metrics: Vec<DemuxedGroupSampleMetrics> =
+        Vec::with_capacity(project_to_metric.len() + 1);
+    let mut project_samples: Vec<SampleMetadata> = Vec::with_capacity(project_to_metric.len() + 1);
+
+    // Sort by project name
+    for key in project_to_metric.keys().sorted() {
+        let metric = project_to_metric.get(key).unwrap();
+        let sample = project_to_sample.get(key).unwrap();
```
Comment (tfenne) [inline src/lib/metrics.rs:279]: Could use `map.remove()` here to take the element out and take ownership so you don't have to clone() below.

**Review by tfenne — APPROVED** [review]
Some minor comments but LGTM!

**Review by omicsorama — APPROVED** [review]

**Pull request merged**